### DISCUSSION TITLE
Skip streamtitle with "adContext="

### DISCUSSION THREAD

**Discussion by skylai82** [schreibfaul1/ESP32-audioI2S#641] [Q&A]
https://stream.rcs.revma.com/crec9cmbv4uvv/4_1nkogcreico9v02/playlist.m3u8
https://stream.rcs.revma.com/crec9cmbv4uvv?playlist.m3u8

https://stream-eu-a.rcs.revma.com/488kt4sbv4uvv/68_1txzsglqw5l3t02/playlist.m3u8
https://stream-eu-a.rcs.revma.com/488kt4sbv4uvv?playlist.m3u8

Hi Folk,
Any idea to skip "adContext=" from https://stream.rcs.revma.com & https://stream-eu-a.rcs.revma.com ? 

Below is my temporary solution.

// 全局变量
const char *streamTitle = NULL;
const char *streamURL = NULL;

void audio_lasthost(const char *url) {
    Serial.print("lasthost    ");
    Serial.println(url);
    // 更新全局变量
    streamURL = url;
}

void audio_showstreamtitle(const char *title) {
    Serial.print("streamtitle ");
    Serial.println(title);
    // 当streamtitle检测是否包含 "adContext="
    if (strstr(title, "adContext=") != NULL) {
        // 包含 "adContext"，执行连接操作
        Serial.print("lasthost::::    ");
        Serial.println(streamURL);
        audio.connecttohost(streamURL);
        delay(1000);
        audio.loop();
        Serial.println("检测到广告");
    }
}

**Comment by skylai82** [comment]
After 2nd time adContext, skip and continue playing audio.loop().

// 全局变量
const char *streamTitle = NULL;
const char *streamURL = NULL;
int adContextCounter = 0;

void audio_showstreamtitle(const char *title) {
    Serial.print("streamtitle ");
    Serial.println(title);
    // 当streamtitle检测是否包含 "adContext="
    if (strstr(title, "adContext=") != NULL && adContextCounter < 2) {
        // 包含 "adContext"，执行连接操作
        Serial.println("检测到广告");
        Serial.print("lasthost::::    ");
        Serial.println(streamURL);
        audio.connecttohost(radioChannels[currentChannel]);
        audio.connecttohost(streamURL);
        // 增加计数器
        adContextCounter++;
    }
}

**Comment by schreibfaul1** [comment]
the parent URL is:
http://n15a-eu.rcs.revma.com/crec9cmbv4uvv/hls.m3u8

**Reply by skylai82** [reply]
Thanks!
But not fully solved, streamtitle no show adContext👍, after few hours play link again, stream audio still play adContext for 1 time😑.

https://radio-online.my/lite <-if stream thru website totally no play adContext 
![image](https://github.com/schreibfaul1/ESP32-audioI2S/assets/37041202/5f956172-21d3-41e8-956c-8db0e341df63)

**Reply by skylai82** [reply]
![image](https://github.com/schreibfaul1/ESP32-audioI2S/assets/37041202/d2e3a04b-48ba-46e7-b9be-c131feb25bfb)
Found some bugs!
If the stream parent URL, when the Server changes from no1 to n13,n14..., the stream link makes no sound.

**Reply by schreibfaul1** [reply]
http://n01a-eu.rcs.revma.com/crec9cmbv4uvv/hls.m3u8
http://n05a-eu.rcs.revma.com/bn4ex8sbv4uvv/hls.m3u8
http://n13a-eu.rcs.revma.com/bn4ex8sbv4uvv/hls.m3u8
http://n14a-eu.rcs.revma.com/bn4ex8sbv4uvv/hls.m3u8

**Reply by skylai82** [reply]
after few hours above link no work already.
must be use http**s**.
https://n01a-eu.rcs.revma.com/crec9cmbv4uvv/hls.m3u8

Second bugs after few hours stream link break(no sound)
?hls.m3u8 instead of /hls.m3u8， 
https://n01a-eu.rcs.revma.com/crec9cmbv4uvv?hls.m3u8

const char *streamURL = "Default_Blank";
void audio_lasthost(const char *url){  //stream URL played
    Serial.print("lasthost:");
    Serial.println(url);
    // 更新全局变量
    streamURL = url;
}
void loop(){
if (Serial.available()) {
        Serial.println("Change Radio");
        String r = Serial.readString(); // 读取Serial数据到inputString
        r.trim(); // 去除空格
        Serial.println(r.c_str());
        audio.connecttohost(r.c_str());
        audio.loop();
    }

if (audio.isRunning()) {
    audio.loop(); //Playing...
    //Serial.println("Playing...");
  } else {
    //audio.stopSong();
    //delay(100);
    Serial.println("Stop playing");
    audio.connecttohost(streamURL);
    audio.loop();
    }
}

**Comment by skylai82** [comment]
> the parent URL is: http://n15a-eu.rcs.revma.com/crec9cmbv4uvv/hls.m3u8


![image](https://github.com/schreibfaul1/ESP32-audioI2S/assets/37041202/477c64a4-a725-4970-b46d-7893695a0f11)

I think I found the solution to Skip streamtitle with "adContext="  & Fast stream.
Thanks alot.
void setup() {
。。。。
。。。。
//First stream Parent link
audio.connecttohost("http://stream.rcs.revma.com/bn4ex8sbv4uvv/hls.m3u8");
audio.loop();
//Second stream Child link
audio.connecttohost("http://stream.rcs.revma.com/bn4ex8sbv4uvv?playlist.m3u8");
audio.loop();
}

void loop(){
audio.loop();
}
